### PR TITLE
rustc_lint: handle more method calls in unconditional_recursion.

### PR DESCRIPTION
**Pull request by eddyb**
After #26694, the overloaded operator and "impl not known at method lookup time" cases started triggering the lint.
I've also added checks for overloaded autoderef and method calls via paths (i.e. `T::method()`).
All new 8 test cases did not trigger the lint before #26694.
r? @huonw

### CODE REVIEW

**Comment by Gankro** [comment]
ping @huonw (also needs rebase)

File: src/librustc_trans/trans/callee.rs [hunk 1]
```diff
@@ -160,16 +160,26 @@ fn trans<'blk, 'tcx>(bcx: Block<'blk, 'tcx>, expr: &ast::Expr)
...
             }
-            def::DefMethod(meth_did, def::FromTrait(trait_did)) => {
-                fn_callee(bcx, meth::trans_static_method_callee(bcx.ccx(),
-                                                                meth_did,
-                                                                trait_did,
-                                                                ref_expr.id,
-                                                                bcx.fcx.param_substs))
+            def::DefMethod(meth_did) => {
+                let method_item = bcx.tcx().impl_or_trait_item(meth_did);
+                fn_callee(bcx, match method_item.container() {
```
Comment (huonw) [inline src/librustc_trans/trans/callee.rs]: Could you move the match to a temporary variable (i.e. to not be inline in the function call)?

**Comment by huonw** [comment]
r=me with a rebase and the nit.

**Pull request merged**